File: insertSort.cpp
#include <iostream>
#include <vector>

using namespace std;

/**************
--------------------------
Insert Sort: O(n*n) stable
--------------------------
Thoght: 
1. iterate an unsorted array but start from the second element since we suppose the 
   first element is ordered.
2. store current element into key and assign the second iterator j which is i-1.
3. use while loop to judge if j is larger or equal to 0 to avoid out of index and arr[j] 
   should be larger than key which means we should do something.
4. After checking the former element and update the current element, we update the former
   element in the end.
5. Note: we assume the former sub-array is ordered already because we handle from the first data
         so that we actually just jump out from the while loop when we checked only once
**************/

void insertSort(vector<int>& arr) {
	for (int i = 1; i < arr.size(); ++i) {
		int key = arr[i];
		int j = i - 1;
		while (j >= 0 && arr[j] > key) {
			arr[j + 1] = arr[j];
			--j;
		}
		arr[j + 1] = key;
	}
}

int main() {
	int num;
	vector<int> nums;
	while (cin >> num) {
		nums.push_back(num);
		if (cin.get() == '\n') {
			break;
		}
	}
	for (int i = 0; i < nums.size(); ++i) {
		cout << nums[i] << ' ';
	}
	cout << endl;
	insertSort(nums);
	for (int i = 0; i < nums.size(); i++) {
		cout << nums[i] << ' ';
	}
	return 0;
}

File: insertSort.js
// best:O(n) worst:O(n^2) avg:O(n^2)

function insertSort(arr){
    // store the length of the array
    const len = arr.length;
    // temp is used to store current element which should be interted
    let temp;
    // i is used to mark the data which should be inserted
    for(let i = 0; i < len; ++i){
        // j is used to help temp find its position
        // from rear to the former
        let j = i;
        temp = arr[i];
        // judge j's former element if it is bigger than temp
        while(j > 0 && arr[j-1] > temp){
            // if true, j-1 moves back one position to make space for temp
            arr[j] = arr[j-1];
            j--;
        }
        // make space in the loop, so j is the correct index of temp in the end
        arr[j] = temp;
    }
    return arr;
}